File: 2023/haskell/src/Day05.hs
module Day05 (part1, part2) where

import qualified Data.List as L
import Data.List.Split (chunksOf)
import qualified Data.Map as M
import qualified Data.Maybe as Maybe
import qualified Data.Set.Internal as S

data Kind
  = Seed
  | Soil
  | Fertilizer
  | Water
  | Light
  | Temperature
  | Humidity
  | Location
  deriving (Show, Eq, Ord)

data MapperEntry = MapperEntry {dest :: Int, src :: Int, len :: Int} deriving (Show, Eq, Ord)

-- data MapperSection = MapperSection {source :: Kind, destination :: Kind, entries :: [MapperEntry]} deriving (Show)

newtype SeedNumber = SeedNumber Int deriving (Show)

data Almanac = Almanac [Int] (M.Map (Kind, Kind) [MapperEntry]) deriving (Show)

getKind :: String -> Kind
getKind "seed" = Seed
getKind "soil" = Soil
getKind "fertilizer" = Fertilizer
getKind "water" = Water
getKind "light" = Light
getKind "temperature" = Temperature
getKind "humidity" = Humidity
getKind "location" = Location
getKind _ = error "Unexpected Kind"

createSection :: [String] -> Kind -> Kind -> M.Map (Kind, Kind) [MapperEntry] -> M.Map (Kind, Kind) [MapperEntry]
createSection es source destination
  = M.insert (source, destination) entries
  where
    -- map

    entries =
      map
        ( ( \ws ->
              MapperEntry
                { dest = (read :: String -> Int) $ head ws,
                  src = (read :: String -> Int) $ ws !! 1,
                  len = (read :: String -> Int) $ ws !! 2
                }
          )
            . words
        )
        es

createAlmanac :: String -> Almanac
createAlmanac input =
  Almanac seeds map7
  where
    m = M.empty
    seedsStr : _ : rest = lines input
    _ : seedsTail = words seedsStr
    seeds = map (read :: String -> Int) seedsTail
    (_ : seeds2Soil, _ : rest1) = span (/= "") rest
    map1 = createSection seeds2Soil Seed Soil m
    (_ : soil2Fertilizer, _ : rest2) = span (/= "") rest1
    map2 = createSection soil2Fertilizer Soil Fertilizer map1
    (_ : fertilizer2Water, _ : rest3) = span (/= "") rest2
    map3 = createSection fertilizer2Water Fertilizer Water map2
    (_ : water2Light, _ : rest4) = span (/= "") rest3
    map4 = createSection water2Light Water Light map3
    (_ : light2Temparature, _ : rest5) = span (/= "") rest4
    map5 = createSection light2Temparature Light Temperature map4
    (_ : temperature2Humidity, _ : rest6) = span (/= "") rest5
    map6 = createSection temperature2Humidity Temperature Humidity map5
    (_ : humidity2Location, _) = span (/= "") rest6
    map7 = createSection humidity2Location Humidity Location map6

nextKind :: Kind -> Maybe Kind
nextKind Seed = Just Soil
nextKind Soil = Just Fertilizer
nextKind Fertilizer = Just Water
nextKind Water = Just Light
nextKind Light = Just Temperature
nextKind Temperature = Just Humidity
nextKind Humidity = Just Location
nextKind Location = Nothing

hasMapping :: MapperEntry -> Int -> Maybe Int
hasMapping MapperEntry {dest = d, src = s, len = l} value =
  if value < s + l && value >= s
    then Just $ d + (value - s)
    else Nothing

findLocationBySeed :: M.Map (Kind, Kind) [MapperEntry] -> Kind -> Int -> Int
findLocationBySeed m sk s = do
  case nextKind sk of
    Just dk ->
      let entries = m M.! (sk, dk)
       in case dropWhile (== Nothing) $ map (`hasMapping` s) entries of
            Just d : _ -> findLocationBySeed m dk d
            _ -> findLocationBySeed m dk s
    Nothing ->
      s

hasMappingRange1 :: MapperEntry -> (Int, Int) -> [(Int, Int)]
hasMappingRange1 entry@MapperEntry {dest = d, src = s, len = l} (v, vl) =
  if v + vl <= s || v >= s + l
    then []
    else case (v < s, v + vl > s + l) of
      (True, True) -> [(v, s - v), (d, l), (s + l, v + vl - (s + l))]
      (True, False) -> [(v, s - v), (d, v + vl - s)]
      (False, True) -> [(d + (v - s), s + l - v), (s + l, (v + vl) - (s + l))]
      (False, False) -> [(d + (v - s), vl)]

hasMappingRange :: Kind -> MapperEntry -> (Int, Int) -> [(Int, Int)]
hasMappingRange _ entry@MapperEntry {dest = d, src = s, len = l} (v, vl) =
  if v + vl <= s || v >= s + l
    then []
    else 

      case (v < s, v + vl > s + l) of
        (True, True) -> [(v, s - v), (d, l), (s + l, v + vl - (s + l))]
        (True, False) -> [(v, s - v), (d, v + vl - s)]
        (False, True) -> [(d + (v - s), s + l - v), (s + l, (v + vl) - (s + l))]
        (False, False) -> [(d + (v - s), vl)]

findLocationBySeedRange :: M.Map (Kind, Kind) [MapperEntry] -> Kind -> [(Int, Int)] -> IO Int

findLocationBySeedRange m sk rs = do
  case nextKind sk of
    Just dk -> do
      case concatMap (filter (not . all null) . (\entry -> map (hasMappingRange sk entry) rs)) (m M.! (sk, dk)) of
        [] -> findLocationBySeedRange m dk rs
        rss ->
          do
            
            return
            $ minimum
            $ mapM
              ( \rs1 ->
                  findLocationBySeedRange m dk $
                    concatMap (\(pair, old) -> if null pair then old else pair) $
                      zip rs1 $
                        map (: []) rs
              )
            $ concat rss


    Nothing ->
      return $ fst $ minimum rs
findLocationBySeedRange _ _ _ = error "Chunks invalid!"

part1 :: String -> IO ()
part1 input = do
  let almanac@(Almanac seeds m) = createAlmanac input
  print $ minimum $ map (findLocationBySeed m Seed) seeds

part2 :: String -> IO ()
part2 input = do
  let almanac@(Almanac seeds m) = createAlmanac input
  let rs = map (\xs -> (head xs, xs !! 1 - 1)) $ chunksOf 2 seeds

  res <- findLocationBySeedRange m Seed rs

  print res

File: 2023/scala/src/main/scala/day05.scala
package day05

import locations.Directory.currentDir
import inputs.Input.loadFileSync
import scala.compiletime.ops.int
import scala.compiletime.ops.long

@main def part1: Unit =
  println(s"The solution is ${part1(loadInput())}")

@main def part2: Unit =
  println(s"The solution is ${part2(loadInput())}")

def loadInput(): String = loadFileSync(
  s"$currentDir/../../../../input/day05"
)

enum Kind:
  case Seed, Soil, Fertilizer, Water, Light, Temparature, Humidity, Location

enum Pair:
  case Mapped(start: Long, len: Long)
  case Unmapped(start: Long, len: Long)

  def value() =
    this match {
      case Mapped(start, len)   => (start, len)
      case Unmapped(start, len) => (start, len)
    }

case class MapInfo(dest: Long, src: Long, len: Long):

  def findDestination(value: Long) =
    if value < src + len && value >= src
    then Some(dest + (value - src))
    else None

  def findDestination(pair: Pair): Option[List[Pair]] =

    val (v, vl) = pair.value()

    if v + vl <= src || v >= src + len
    then None
    else

      val ls = (v < src, v + vl > src + len) match
        case (true, true) =>
          List(
            Pair.Unmapped(v, src - v),
            Pair.Mapped(dest, len),
            Pair.Unmapped(src + len, v + vl - (src + len))
          )
        case (true, false) =>
          List(Pair.Unmapped(v, src - v), Pair.Mapped(dest, v + vl - src))
        case (false, true) =>
          List(
            Pair.Mapped(dest + (v - src), src + len - v),
            Pair.Unmapped(src + len, (v + vl) - (src + len))
          )
        case (false, false) =>
          List(Pair.Mapped(dest + (v - src), vl))

      Some(ls)

case class Almanac(
    seeds: Array[Long],
    mappers: Map[(Kind, Kind), List[MapInfo]]
):
  def findLocationBySeed(sk: Kind, s: Long): Long =
    nextKind(sk) match {
      case Some(dk) =>
        val entries = mappers((sk, dk))
        val x = entries
          .map(mi => mi.findDestination(s))
          .dropWhile(_.isEmpty)
          .toList match {
          case Some(d) :: _ => findLocationBySeed(dk, d)
          case _            => findLocationBySeed(dk, s)
        }
        x
      case None => s
    }

  def findLocationBySeedRange(sk: Kind, pairs: List[Pair]): Long =
    nextKind(sk) match {
      case Some(dk) =>
        val entries = mappers((sk, dk))

        val lss: List[(Option[List[day05.Pair]], day05.Pair)] = entries
          .flatMap(mi => pairs.map(pair => (mi.findDestination(pair), pair)))

        if lss.forall(_._1.isEmpty) then findLocationBySeedRange(dk, pairs)
        else

          val emptyDefaults = lss
            .filter(_._1.isEmpty)
            .map((_, p) => Pair.Unmapped(p.value()._1, p.value()._2))
            .toSet
            .toList

          val (mapped, unmapped) =
            lss.filter(!_._1.isEmpty).map(_._1.get).flatten.partition {
              case (Pair.Mapped(start, len)) => true
              case Pair.Unmapped(start, len) => false
            }

          val unmapped1 = Set.newBuilder[Pair]

          for uls <- pairs do {
            val (us, ul) = uls.value()

            if mapped.isEmpty then unmapped1 += Pair.Unmapped(us, ul)
            else

              for mls <- mapped do {
                val (ms, ml) = mls.value()

                if !(us + ul <= ms || us >= ms + ml) then {

                  (us <= ms, ms + ml <= us + ul) match {

                    case (true, true) =>
                      if ms - us > 0 then
                        unmapped1 += Pair.Unmapped(us, ms - us)

                      if us + ul - (ms + ml) > 0 then
                        unmapped1 += Pair.Unmapped(ms + ml, us + ul - (ms + ml))

                    case (true, false) =>
                      if ms - us > 0 then
                        unmapped1 += Pair.Unmapped(us, ms - us)

                    case (false, true) =>
                      if us + ul - (ms + ml) > 0 then
                        unmapped1 += Pair.Unmapped(ms + ml, us + ul - (ms + ml))

                    case (false, false) => unmapped1 += Pair.Unmapped(us, ul)

                  }

                }

              }

          }

          val ms = mapped.toSet.toList
          val us = unmapped1.result().toList

          val xs = ms ++ us
          findLocationBySeedRange(dk, xs)

      case None =>
        pairs.map(p => p.value()._1).min
    }

def createAlmanac(input: String) =
  input.split("\n\n") match
    case Array(
          seeds_raw,
          seed2soil,
          soil2fertilizer,
          fertilizer2water,
          water2light,
          light2temp,
          tem2hum,
          hum2loc
        ) =>
      def getEntries(s: String) = {
        s.split("\n")
          .toList
          .drop(1)
          .map { case s"$dest $src $len" =>
            MapInfo(dest.toLong, src.toLong, len.toLong)
          }
          .toList
      }

      var mappers: Map[(Kind, Kind), List[MapInfo]] = Map()
      val seeds = seeds_raw match {
        case s"seeds: $rest" =>
          rest.split(" ").map(_.toLong)
      }

      mappers += ((Kind.Seed, Kind.Soil) -> getEntries(seed2soil))
      mappers += ((Kind.Soil, Kind.Fertilizer) -> getEntries(soil2fertilizer))
      mappers += ((Kind.Fertilizer, Kind.Water) -> getEntries(fertilizer2water))
      mappers += ((Kind.Water, Kind.Light) -> getEntries(water2light))
      mappers += ((Kind.Light, Kind.Temparature) -> getEntries(light2temp))
      mappers += ((Kind.Temparature, Kind.Humidity) -> getEntries(tem2hum))
      mappers += ((Kind.Humidity, Kind.Location) -> getEntries(hum2loc))

      Almanac(seeds, mappers)

def nextKind(k: Kind) =
  k match
    case Kind.Seed        => Some(Kind.Soil)
    case Kind.Soil        => Some(Kind.Fertilizer)
    case Kind.Fertilizer  => Some(Kind.Water)
    case Kind.Water       => Some(Kind.Light)
    case Kind.Light       => Some(Kind.Temparature)
    case Kind.Temparature => Some(Kind.Humidity)
    case Kind.Humidity    => Some(Kind.Location)
    case Kind.Location    => None

def part1(input: String): String =
  val almanac = createAlmanac(input)
  almanac.seeds
    .map(s => almanac.findLocationBySeed(Kind.Seed, s))
    .min
    .toString()

def part2(input: String): String =

  val almanac = createAlmanac(input)
  // val pairs =
  //   almanac.seeds.grouped(2).map(a => Pair.Unmapped(a(0), a(1))).toList

  val pairs =
    almanac.seeds
      .grouped(2)
      .map(a => Pair.Unmapped(a(0), a(0) + a(1) - 1))
      .toList
  
  val x = almanac.findLocationBySeedRange(Kind.Seed, pairs)
  
  x.toString()

@main def part12: Unit =

  val input = loadInput()
